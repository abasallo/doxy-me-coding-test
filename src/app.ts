import * as dotenv from 'dotenv'

import express from 'express'
import cors from 'cors'
import bodyParser from 'body-parser'

import index from './express/routes'
import user from './express/routes/user'
import product from './express/routes/product'
import review from './express/routes/review'

import { Container } from './container'

dotenv.config()

export const app = express()

app.use(cors())

app.use(bodyParser.urlencoded({ extended: false }))
app.use(bodyParser.json())

app.use('/', index)
app.use('/user', user)
app.use('/product', product)
app.use('/review', review)

app.use(Container.Middlewares.error)
app.use(Container.Middlewares.log)

export default app
