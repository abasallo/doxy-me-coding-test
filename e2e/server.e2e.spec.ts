import axios from 'axios'

import { ReasonPhrases, StatusCodes } from 'http-status-codes'

import app from '../src/app'

import http from 'http'

const TEST_SERVER_PORT = 4001

let server: http.Server

let token: string

beforeAll(async () => {
  server = await app.listen({ port: TEST_SERVER_PORT }, () => console.log(`Test Server initialised on port: ${TEST_SERVER_PORT}`))
})

afterAll(() => {
  server.close()
})

describe('Server E2E tests', () => {
  describe('User E2E tests', () => {
    test('Sign up', async () => {
      const { status, statusText, data } = await axios({
        method: 'post',
        url: `http://localhost:${TEST_SERVER_PORT}/user/sign-up/`,
        data: {
          username: 'username-test',
          password: 'password-test'
        }
      })

      expect(status).toEqual(StatusCodes.OK)
      expect(statusText).toEqual(ReasonPhrases.OK)

      expect(data).toMatchSnapshot({
        id: expect.any(Number)
      })
    })

    test('Sign in', async () => {
      const { status, statusText, data } = await axios({
        method: 'post',
        url: `http://localhost:${TEST_SERVER_PORT}/user/sign-in/`,
        data: {
          username: 'username-test',
          password: 'password-test'
        }
      })

      expect(status).toEqual(StatusCodes.OK)
      expect(statusText).toEqual(ReasonPhrases.OK)

      token = data.token

      expect(data).toMatchSnapshot({
        token: expect.any(String)
      })
    })
  })

  describe('Product E2E tests', () => {
    let newProductId: string

    test('Add Product', async () => {
      const { status, statusText, data } = await axios({
        method: 'post',
        url: `http://localhost:${TEST_SERVER_PORT}/product/`,
        data: {
          description: 'new-description'
        },
        headers: { token }
      })

      expect(status).toEqual(StatusCodes.OK)
      expect(statusText).toEqual(ReasonPhrases.OK)

      newProductId = data.id

      expect(data).toMatchSnapshot({
        id: expect.any(Number),
        userId: expect.any(Number)
      })
    })

    test('Find Products', async () => {
      const { status, statusText, data } = await axios({
        method: 'get',
        url: `http://localhost:${TEST_SERVER_PORT}/product/find`,
        data: {
          userId: 3
        }
      })

      expect(status).toEqual(StatusCodes.OK)
      expect(statusText).toEqual(ReasonPhrases.OK)

      expect(data).toMatchSnapshot()
    })

    test('Get Product', async () => {
      const { status, statusText, data } = await axios({
        method: 'get',
        url: `http://localhost:${TEST_SERVER_PORT}/product/${newProductId}`
      })

      expect(status).toEqual(StatusCodes.OK)
      expect(statusText).toEqual(ReasonPhrases.OK)

      expect(data).toMatchSnapshot({
        id: expect.any(Number),
        userId: expect.any(Number)
      })
    })

    test('Update Product', async () => {
      const { status, statusText, data } = await axios({
        method: 'put',
        url: `http://localhost:${TEST_SERVER_PORT}/product/`,
        data: {
          id: newProductId,
          description: 'updated-description'
        },
        headers: { token }
      })

      expect(status).toEqual(StatusCodes.OK)
      expect(statusText).toEqual(ReasonPhrases.OK)

      expect(data).toMatchSnapshot({
        id: expect.any(Number),
        userId: expect.any(Number)
      })
    })

    test('Delete Product', async () => {
      const { status, statusText, data } = await axios({
        method: 'delete',
        url: `http://localhost:${TEST_SERVER_PORT}/product/${newProductId}`,
        headers: { token }
      })

      expect(status).toEqual(StatusCodes.OK)
      expect(statusText).toEqual(ReasonPhrases.OK)

      expect(data).toMatchSnapshot({
        id: expect.any(Number),
        userId: expect.any(Number)
      })
    })
  })

  describe('Review E2E tests', () => {
    test('Find Reviews', async () => {
      const { status, statusText, data } = await axios({
        method: 'get',
        url: `http://localhost:${TEST_SERVER_PORT}/review/find`,
        headers: { token },
        data: {
          productId: 3
        }
      })

      expect(status).toEqual(StatusCodes.OK)
      expect(statusText).toEqual(ReasonPhrases.OK)

      expect(data).toMatchSnapshot()
    })

    test('Get Review', async () => {
      const { status, statusText, data } = await axios({
        method: 'get',
        url: `http://localhost:${TEST_SERVER_PORT}/review/1`,
        headers: { token }
      })

      expect(status).toEqual(StatusCodes.OK)
      expect(statusText).toEqual(ReasonPhrases.OK)

      expect(data).toMatchSnapshot({
        id: expect.any(Number),
        productId: expect.any(Number),
        userId: expect.any(Number)
      })
    })

    test('Add Review', async () => {
      const { status, statusText, data } = await axios({
        method: 'post',
        url: `http://localhost:${TEST_SERVER_PORT}/review/`,
        data: {
          productId: 1,
          text: 'Review Text'
        },
        headers: { token }
      })

      expect(status).toEqual(StatusCodes.OK)
      expect(statusText).toEqual(ReasonPhrases.OK)

      expect(data).toMatchSnapshot({
        id: expect.any(Number),
        productId: expect.any(Number),
        userId: expect.any(Number)
      })
    })
  })

  describe('Health Check E2E tests', () => {
    test('Expected response for Health Check', async () => {
      const { status, statusText, data } = await axios({
        method: 'get',
        url: `http://localhost:${TEST_SERVER_PORT}/`
      })

      expect(status).toEqual(StatusCodes.OK)
      expect(statusText).toEqual(ReasonPhrases.OK)

      expect(data).toMatchSnapshot()
    })
  })
})
