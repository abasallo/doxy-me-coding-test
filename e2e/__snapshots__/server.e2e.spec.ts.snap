// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Server E2E tests Health Check E2E tests Expected response for Health Check 1`] = `"OK"`;

exports[`Server E2E tests Product E2E tests Add Product 1`] = `
{
  "description": "new-description",
  "id": Any<Number>,
  "userId": Any<Number>,
}
`;

exports[`Server E2E tests Product E2E tests Delete Product 1`] = `
{
  "description": "updated-description",
  "id": Any<Number>,
  "userId": Any<Number>,
}
`;

exports[`Server E2E tests Product E2E tests Find Products 1`] = `
[
  {
    "description": "description-2",
    "id": 2,
    "userId": 3,
  },
  {
    "description": "description-3",
    "id": 3,
    "userId": 3,
  },
]
`;

exports[`Server E2E tests Product E2E tests Get Product 1`] = `
{
  "description": "new-description",
  "id": Any<Number>,
  "userId": Any<Number>,
}
`;

exports[`Server E2E tests Product E2E tests Update Product 1`] = `
{
  "description": "updated-description",
  "id": Any<Number>,
  "userId": Any<Number>,
}
`;

exports[`Server E2E tests Review E2E tests Add Review 1`] = `
{
  "id": Any<Number>,
  "productId": Any<Number>,
  "text": "Review Text",
  "userId": Any<Number>,
}
`;

exports[`Server E2E tests Review E2E tests Find Reviews 1`] = `
[
  {
    "id": 2,
    "productId": 3,
    "text": "review-text-2",
    "userId": 2,
  },
  {
    "id": 3,
    "productId": 3,
    "text": "review-text-3",
    "userId": 3,
  },
]
`;

exports[`Server E2E tests Review E2E tests Get Review 1`] = `
{
  "id": Any<Number>,
  "productId": Any<Number>,
  "text": "review-text-1",
  "userId": Any<Number>,
}
`;

exports[`Server E2E tests User E2E tests Sign in 1`] = `
{
  "token": Any<String>,
}
`;

exports[`Server E2E tests User E2E tests Sign up 1`] = `
{
  "id": Any<Number>,
  "username": "username-test",
}
`;
